### PR TITLE
Account for max token length in split func

### PR DESCRIPTION
**Pull request by kevwang**
Build a wrapper over `*bufio.Scanner` which holds new fields.

While `maxTokenLength` could be implemented without being a field, `currIsSeparator` is stateful, so we should just add them together. `currIsSeparator` logic to be added as a follow-up

### CODE REVIEW

File: words/scanner_test.go [hunk 1]
```diff
@@ -381,3 +380,33 @@ func BenchmarkUnicodeSegments(b *testing.B) {
 		b.ReportMetric(float64(c), "tokens")
 	}
 }
+
+func TestScannerMaxTokenLen(t *testing.T) {
+	// Build a string longer than the max token length
+	str := ""
+	for i := 0; i < words.MaxTokenLengthDefault + 5; i++ {
+		if i == words.MaxTokenLengthDefault {
+			str += "."
```
Comment (markdanielfs) [inline words/scanner_test.go]: Looks like this test checks that we cleanly tokenize if the "." comes right after the default length, but what is the behavior if it's after that? Like if we have (a*255)aaa.aaa (where there are more initial a's than max default). Maybe add a test to show that behavior.

**Comment by markdanielfs** [comment]
Added a comment to add a test to help me understand the behavior, but looks good other than that.

File: words/scanner_test.go (hunk 1 above)
Comment (kevwang) [reply words/scanner_test.go]: Sg, just updated and added another test

File: words/scanner_test.go [hunk 2]
```diff
@@ -381,3 +380,47 @@ func BenchmarkUnicodeSegments(b *testing.B) {
...
+	str := ""
+	for i := 0; i < words.MaxTokenLengthDefault; i++ {
+		str += "a"
+	}
+	testCases := []struct{
+		input string
+		expected []string
+	}{
+		{
+			input: str + ".aaa",
```
Comment (claygoddard) [inline words/scanner_test.go:395]: What if we use `str[:len(str)-1] + ".aaa"`? What is supposed to happen? I forget.

File: words/scanner.go [hunk 3]
```diff
@@ -143,6 +171,9 @@ func SplitFunc(data []byte, atEOF bool) (advance int, token []byte, err error) {
 					w = w2
 
 					pos += w
```
Comment (claygoddard) [inline words/scanner.go:173]: Is it possible in all of these different cases that `pos += w` actually takes us over the limit?

File: words/scanner_test.go (hunk 2 above)
Comment (kevwang) [reply words/scanner_test.go:395]: Hmm, let me see what Solr would do.

File: words/scanner.go (hunk 3 above)
Comment (kevwang) [reply words/scanner.go:173]: Yes that's possible, going to see how Solr handles that

File: words/scanner_test.go (hunk 2 above)
Comment (kevwang) [reply words/scanner_test.go:395]: Solr will automatically strip the `"."`, whereas this implementation will keep it at the end. Will modify

File: words/scanner.go (hunk 3 above)
Comment (kevwang) [reply words/scanner.go:173]: Solr doesn't care about the byte size, so I'll change this to increment a new var instead

**Pull request closed without merging**